### DISCUSSION TITLE
Avoiding computation of initial values on store creation

### DISCUSSION THREAD

**Discussion by silverwind** [pmndrs/zustand#717] [Q&A]
I have a persisted store with a variable whose initial value is the result of a function:

```js
import create from "zustand";
import {persist} from "zustand/middleware";

export const useStore = create(persist(set => ({
  uuid: crypto.randomUUID(),
}), {name: "store"}));
```

By the nature of JavaScript, `crypto.randomUUID()` is called every time this code runs. Is it possible to avoid such useless and expensive computations during store initialization?

I thought about doing `() => crypto.randomUUID()` but I think such a function variable won't work as expected.

**Comment by dai-shi** [comment]
Is it specific to `persist`?

One thing I can think of is to initialize with `null` and run the code afterward.

```js
export const useStore = create(set => ({
  uuid: null,
}));

const initializeStore = () => {
  useStore.setStore({
    uuid: crypto.randomUUID(),
  })
}
```

**Comment by silverwind** [comment]
It's kind of specific to `persist` as I'd like to skip the initialization only when `persist` did restore from storage.

More generally, I think `zustand` lacks a way to "lazily" initialize a store variable, e.g. similar to React's [`useState(fn)`](https://reactjs.org/docs/hooks-reference.html#lazy-initial-state) which `persist` could detect and not call in case it's restoring from store (and by doing that, disregard any initial value anyways).

**Reply by dai-shi** [reply]
I think it's specific to persist and only to sync storage. zustand store's initialization is done on creation, so there's no notion of lazy initialization like useState.

I think you could use `merge` option in `persist` to initialize some values lazily.
